federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,885 @@
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		err = clientset.Core().Services(service.Namespace).Delete(service.Name, &api.DeleteOptions{})
+		if err == nil || errors.IsNotFound(err) {
+			glog.V(4).Infof("service %s/%s deleted from cluster %s", service.Namespace, service.Name, clusterName)
+			return nil, !retryable
+		}
+		time.Sleep(cachedService.nextRetryDelay())
+	}
+	glog.V(4).Infof("failed to delete service %s/%s from cluster %s, %+v", service.Namespace, service.Name, clusterName, err)
+	return err, retryable
+}
+
+func (s *ServiceController) ensureClusterService(cachedService *cachedService, clusterName string, service *api.Service, client *clientset.Clientset) error {
+	var err error
+	var needUpdate bool
+	for i := 0; i < clientRetryCount; i++ {
+		svc, err := client.Core().Services(service.Namespace).Get(service.Name)
+		if err == nil {
+			// service exists
+			glog.V(5).Infof("found service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+			//reserve immutable fields
+			service.Spec.ClusterIP = svc.Spec.ClusterIP
+
+			//reserve auto assigned field
+			for i, oldPort := range svc.Spec.Ports {
+				for _, port := range service.Spec.Ports {
+					if port.NodePort == 0 {
+						if !portEqualExcludeNodePort(&oldPort, &port) {
+							svc.Spec.Ports[i] = port
+							needUpdate = true
+						}
+					} else {
+						if !portEqualForLB(&oldPort, &port) {
+							svc.Spec.Ports[i] = port
+							needUpdate = true
+						}
+					}
+				}
+			}
+
+			if needUpdate {
+				// we only apply spec update for now
+				svc.Spec = service.Spec
+				_, err = client.Core().Services(svc.Namespace).Update(svc)
+				if err == nil {
+					glog.V(5).Infof("service %s/%s successfully updated to cluster %s", svc.Namespace, svc.Name, clusterName)
+					return nil
+				} else {
+					glog.V(4).Infof("failed to update %+v", err)
+				}
+			} else {
+				glog.V(5).Infof("service %s/%s is not updated to cluster %s as the spec are identical", svc.Namespace, svc.Name, clusterName)
+				return nil
+			}
+		} else if errors.IsNotFound(err) {
+			// Create service if it is not found
+			glog.Infof("Service '%s/%s' is not found in cluster %s, trying to create new",
+				service.Namespace, service.Name, clusterName)
+			service.ResourceVersion = ""
+			_, err = client.Core().Services(service.Namespace).Create(service)
+			if err == nil {
+				glog.V(5).Infof("service %s/%s successfully created to cluster %s", service.Namespace, service.Name, clusterName)
+				return nil
+			}
+			glog.V(4).Infof("failed to create %+v", err)
+			if errors.IsAlreadyExists(err) {
+				glog.V(5).Infof("service %s/%s already exists in cluster %s", service.Namespace, service.Name, clusterName)
+				return nil
+			}
+		}
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
```

Comment:
nit: Some infof start with a capital letter, some dont :)
Will be good to be consistent.